README.md
# Pi Status Board (psb)

![GitHub Workflow Status](https://img.shields.io/github/workflow/status/rj175/pi-status-board/Python%20package)
![GitHub issues](https://img.shields.io/github/issues/rj175/pi-status-board)
![GitHub](https://img.shields.io/github/license/rj175/pi-status-board)
![PyPI - Python Version](https://img.shields.io/pypi/pyversions/psb)
![PyPI - Downloads](https://img.shields.io/pypi/dm/psb)

The Pi Status Board (psb) can be used to display text or pre-defined images on an eInk display attached to a Raspberry Pi. It is controlled via AWS IoT Core and uses mqtt.

## Requirements

Psb has been designed to work on a Raspberry Pi using and an [Inky pHAT display.](https://shop.pimoroni.com/products/inky-phat)

## Install

Use package manager `pip` or `pipenv` to install psb:

```bash
pip install psb
```

## Usage

Psb is a command line utility. The configuration file should be stored in `/etc/psb/conf.ini`.  A sample config file can be found within the [resources folder](resources/conf.ini).

For a full breakdown of all configuration options please see the [configuration](docs/configuration.md) documentation.

To run psb, run the following command without any arguments:

```bash
psb
```

Psb will connect to the mqtt broker specified in the configuration file.

To trigger the changing of the display, a message needs to be sent to the configured mqtt topic. There are two types of messages that can be processed.

For a text message the message sent to the topic should be:

```json
{
    "type": "text",
    "data": "This is a message"
}
```

For images, the data field should contain the name of the image to display without the extension:

```json
{
    "type": "image",
    "data": "meeting"
}
```

### Running as a service

If you would like to run psb as a service, copy the example systemd [service file](resources/status-board.serice) to `/lib/systemd/system/pi-status-board.service`. You may need to change the path to the location of psb, this can be found by using the `whereis` command:

```bash
> whereis psb
psb: /etc/psb /home/pi/.local/bin/psb
```

After creating the file, reload the systemd daemon:

```bash
> sudo systemctl daemon-reload
```

You should now be able to see the status of the service by running:

```bash
> sudo systemctl status pi-status-board
```

## Contributing

Pull requests are welcome. For major changes, please open an issue first to discuss what you would like to change.

Please make sure to update tests as appropriate.

## License

Released under GNU General Public License v3.0.


CHANGELOG.md
# Changelog

<a name="0.17.0"></a>
## 0.17.0 (2021-04-28)

### Changed

- 🎨 Updated setup.py with README contents [[7aa3e14](https://github.com/rj175/aws-iot/commit/7aa3e14f77859bd7d57b9b99d72736c11b85a51a)]
- 🚚 Created resources dir [[9b7dfbf](https://github.com/rj175/aws-iot/commit/9b7dfbf34fba8cbf1d55f59ce6a650da80f0593b)]
- 🚚 Moved config and systemd examples [[2a09d16](https://github.com/rj175/aws-iot/commit/2a09d16092a39282b7f1c1d03747076c4f6562bb)]

### Miscellaneous

- 📄 Changed license filename [[6261e89](https://github.com/rj175/aws-iot/commit/6261e89b90240b7e69ee0127591e1ee13fa1c5f1)]
- 📝 Updated README.md [[c75cac2](https://github.com/rj175/aws-iot/commit/c75cac2ae2eb2ea89eda3ae698c54afbe493b242)]
- 📝 Updated README.md [[33de34a](https://github.com/rj175/aws-iot/commit/33de34afb2d32465c39b197b9f7eb1e186c054a7)]
- 📝 Added docs [[1305636](https://github.com/rj175/aws-iot/commit/1305636454b147e753559e667957463ff9d80329)]
- 📄 Added LICENSE.md [[3ce6434](https://github.com/rj175/aws-iot/commit/3ce64344b758a70a1dca50ea8094ed753164b51a)]


<a name="0.16.0"></a>
## 0.16.0 (2021-04-26)

### Changed

- 🎨 Changes to setup.py [[deedc45](https://github.com/rj175/aws-iot/commit/deedc4554438233acb98b08d008db94f5795573b)]


<a name="0.15.0"></a>
## 0.15.0 (2021-04-26)

### Added

- ➕ Added missing rpi.gpio from requirements [[9eba54a](https://github.com/rj175/aws-iot/commit/9eba54a4a12244ab32cb2512a7cae2dbd12d0684)]
- ➕ Added missing requirements.txt dep [[4491a9b](https://github.com/rj175/aws-iot/commit/4491a9bc42b838647a22a003dbef29c630b1481b)]

### Fixed

- 💚 Added missing dep [[21333aa](https://github.com/rj175/aws-iot/commit/21333aaf48b205335998cf38ab7c92240ab3f89b)]

### Miscellaneous

-  Create python-publish.yml [[ced22e6](https://github.com/rj175/aws-iot/commit/ced22e61535362125c92244059f4a49b9f558e99)]


<a name="0.14.0"></a>
## 0.14.0 (2021-04-26)

### Added

- ✅ Addition of tests [[715c9be](https://github.com/rj175/aws-iot/commit/715c9be937a20e6c57556d5032204233e9c1096a)]
- ➕ Dependency changes [[e1a2572](https://github.com/rj175/aws-iot/commit/e1a25724f2e36730ea61b0e76f901e46aae8c9c2)]
- ➕ Updated dependencies [[a8281e5](https://github.com/rj175/aws-iot/commit/a8281e5772e4435345cc78e1c043b68a5d5b9551)]
- ➕ Updated dependencies [[6698f8d](https://github.com/rj175/aws-iot/commit/6698f8daba5b8c4d97ab9728ed9c203541dc67e1)]
- 👷‍♂️ Updated CI config [[afb57b0](https://github.com/rj175/aws-iot/commit/afb57b00587cf579f2f9371bd89916641efad2df)]
- 👷‍♂️ Added linting [[545d003](https://github.com/rj175/aws-iot/commit/545d003756ea74567677d0a5a870b95d5093a806)]
- 👷‍♂️ Update to CI [[68170ea](https://github.com/rj175/aws-iot/commit/68170eab65ea9f7a7ac79865515d75e05939c566)]
- 👷‍♂️ Changes to pytest [[c329c14](https://github.com/rj175/aws-iot/commit/c329c1404c55b341f5b038c3b20bec7f91771d52)]
- 👷‍♂️ Change to pytest [[72467e9](https://github.com/rj175/aws-iot/commit/72467e9cb0485f22ed3bfc320d23b358d5f6fdff)]
- ➕ Added coverage as a dependancy [[b7992c5](https://github.com/rj175/aws-iot/commit/b7992c53b30dad9c00a5429333300788606f6d75)]
- ✅ Added .coveragerc file for tests [[8ab3fc5](https://github.com/rj175/aws-iot/commit/8ab3fc5bb84e8300fb97dd15a329f77b712295d0)]
- 👷‍♂️ Added missing coverage install [[835e8c0](https://github.com/rj175/aws-iot/commit/835e8c029bf3423cd14c6c893884bf5dadf7379f)]
- 👷‍♂️ Added coverage step to CI [[f784ad8](https://github.com/rj175/aws-iot/commit/f784ad88b6371e84833d56f50ff5d8e6261504ea)]
- 👷‍♂️ Added test results to CI [[68e072a](https://github.com/rj175/aws-iot/commit/68e072a6afe2781030537af27de7c2484f47c73b)]
- 👷‍♂️ Added CI [[6800648](https://github.com/rj175/aws-iot/commit/680064896c25ead1c875c2f0834129e33c0beab9)]
- ✅ Added tests [[100966f](https://github.com/rj175/aws-iot/commit/100966f66860f0cb12128ff92927a19f03a5281f)]
- ✨ Added setup.py [[eb487bd](https://github.com/rj175/aws-iot/commit/eb487bdb33709a85d5d50b653d38de44fc6e0de2)]
- ➕ Added dev dependencies [[2c7a8a7](https://github.com/rj175/aws-iot/commit/2c7a8a7e5aeb19e608a5eb3220364646637f8738)]
- ✨ changed font [[9cfc7dd](https://github.com/rj175/aws-iot/commit/9cfc7dd7c94fa39d8e630dd77c09693fb55eae86)]
- ✨ trying multiline text [[25c721d](https://github.com/rj175/aws-iot/commit/25c721d9010ac71121c92bd9c86567b2b923e18b)]
- ✨ added set eink to text [[1f9a9bd](https://github.com/rj175/aws-iot/commit/1f9a9bd8dd253468f5250b7c90bfec84348e20b5)]
- ✨ added extra logging [[05ca58e](https://github.com/rj175/aws-iot/commit/05ca58e0829ed79a2db4451e05e4475a03680abe)]
- ✨ added service file [[2e977b7](https://github.com/rj175/aws-iot/commit/2e977b704d02c38bf5be734ea1043945029ca3a7)]
- ✨ moved certs to &#x27;certs&#x27; dir [[d8c3774](https://github.com/rj175/aws-iot/commit/d8c37746bdc7400ea3a7dca8c637d23387b7e321)]
- ✨ first commit [[3cac994](https://github.com/rj175/aws-iot/commit/3cac9940d7e0f3bd0538302880f75446b44efd50)]

### Changed

- 🎨 Added addition of RPi.GPIO to setup.py [[32fc0c2](https://github.com/rj175/aws-iot/commit/32fc0c23490d2ffcd7d61e06f957c8c4f04b400d)]
- 🎨 Removed config logic from main [[f814db5](https://github.com/rj175/aws-iot/commit/f814db534177c7b954236d343cccbb235007157b)]
- 🎨 Changes to display module [[6a3d86f](https://github.com/rj175/aws-iot/commit/6a3d86fbc960b6a9224aeac0fcb7270ce79e4862)]
- 🎨 Moved config [[b2d0af7](https://github.com/rj175/aws-iot/commit/b2d0af7bae08a56a979ca3eb736f65f209bc95da)]
- 🚨 Pylama warnings fixed [[2433215](https://github.com/rj175/aws-iot/commit/24332159367a5c1f3d23aced02a7b4a7330faf54)]
- 🔧 Updated pylama.ini [[a919d01](https://github.com/rj175/aws-iot/commit/a919d01f5ab598e1dd8934e60a96c357751fed29)]
- 🔧 Added pylama config file [[7b2cf0c](https://github.com/rj175/aws-iot/commit/7b2cf0c744d65f3df7e03c176ed13695cebece14)]
- 🚚 Moved code to new layout [[f7769e2](https://github.com/rj175/aws-iot/commit/f7769e2c6d917a86a4f41f5f35d00536af3d65fa)]

### Removed


tests/test_conf.py
"""Tests for config parsing."""
import pytest
from psb import conf


def test_config_read_file_not_found():
    """Test that a FileNotFoundError is raised if config file does not exist.
    """
    with pytest.raises(FileNotFoundError):
        conf.PsbConfig(config_file='conf-none.ini')


def test_config_read():
    """Test no error when config file exists and is read.
    """
    conf.PsbConfig(config_file="tests/test-config-files/conf.ini")


def test_config_root_ca():
    """Test that the correct value is returned for root CA.
    """
    config = conf.PsbConfig(config_file="tests/test-config-files/conf.ini")
    assert config.root_ca == '/etc/psb/rootCA.crt'


def test_config_iot_cert():
    """Test that the correct value is returned for the IoT cert.
    """
    config = conf.PsbConfig(config_file="tests/test-config-files/conf.ini")
    assert config.aws_iot_cert == '/etc/psb/cert.crt'


def test_config_iot_priv_key():
    """Test that the correct value is returned for the IoT private key.
    """
    config = conf.PsbConfig(config_file="tests/test-config-files/conf.ini")
    assert config.aws_iot_priv_key == '/etc/psb/private.key'


def test_config_endpoint():
    """Test that endpoint value is returned.
    """
    config = conf.PsbConfig(config_file="tests/test-config-files/conf.ini")
    assert config.endpoint == "endpoint.iot.us-east-1.amazonaws.com"


def test_config_port():
    """Test that the port is returned if value is in config.
    """
    config = conf.PsbConfig(config_file="tests/test-config-files/conf-with-port.ini")
    assert config.port == "1234"


def test_config_port_no_websocket():
    """Test that the default port is returned when not using websockets.
    """
    config = conf.PsbConfig(config_file="tests/test-config-files/conf.ini")
    assert config.port == 8883


def test_config_port_websocket():
    """Test that the default port is returned when not using websockets.
    """
    config = conf.PsbConfig(config_file="tests/test-config-files/conf-use-websocket.ini")
    assert config.port == 443


def test_config_default_status():
    """Test that the correct default_status from config is returned.
    """
    config = conf.PsbConfig(config_file="tests/test-config-files/conf.ini")
    assert config.default_status == "idle"


def test_config_img_location():
    """Test that the correct img_location from config is returned.
    """
    config = conf.PsbConfig(config_file="tests/test-config-files/conf.ini")
    assert config.img_location == "/etc/psb/img"


def test_config_client_id():
    """Test that the correct client_id from config is returned.
    """
    config = conf.PsbConfig(config_file="tests/test-config-files/conf.ini")
    assert config.client_id == "test_client_id"


def test_config_incorrect_mode_value_error():
    """Test that a ValueError exception is raise when mode is not a valid value.
    """
    with pytest.raises(ValueError):
        conf.PsbConfig(config_file='tests/test-config-files/conf-incorrect-mode.ini')


def test_config_websockets_and_certs_system_exit():
    """Test that a SystemExit exception is raise when use_websockets is true and certs provided.
    """
    with pytest.raises(SystemExit):
        conf.PsbConfig(config_file='tests/test-config-files/conf-use-websocket-certs-provided.ini')


def test_config_no_certs_system_exit():
    """Test that a SystemExit exception is raise when no certs provided when not using websockets`.
    """
    with pytest.raises(SystemExit):
        conf.PsbConfig(config_file='tests/test-config-files/conf-no-certs.ini')


tests/test_display.py
"""Tests for the display module."""
import pytest
from psb import display


def test_image_display_no_file():
    """Test that a FileNotFoundError is raised when image does not exist.
    """
    test_display = display.EinkDisplay()
    with pytest.raises(FileNotFoundError):
        test_display.image(path='tests/resources/images', status='no-status')


def test_image_display():
    """Test to check that the _eink_display function is run, this does not check that the image was displayed.
    """
    test_display = display.EinkDisplay()
    with pytest.raises(RuntimeError):
        test_display.image(path='tests/resources/images', status='available')


def test_process_message_two_lines_under_20():
    """
    Test for a message with new lines each under 20 characters.
    Lines should return as two parts.
    """
    expected_return = ['Hello', 'There']
    test_display = display.EinkDisplay()
    message = "Hello\nThere"
    return_data = test_display.process_message(message)
    assert return_data == expected_return


def test_process_message_long_message():
    """Test for a long message to make sure it is processed properly.
    """
    expected_return = ['Hello There', 'Long Test Message To', 'Split']
    test_display = display.EinkDisplay()
    message = "Hello There\nLong Test Message To Split"
    return_data = test_display.process_message(message)
    print(return_data)
    assert return_data == expected_return


sub.py
from AWSIoTPythonSDK.MQTTLib import AWSIoTMQTTClient
import logging
import time
import json
import configparser
import os
from inky import BLACK
from inky.auto import auto
from font_fredoka_one import FredokaOne
from PIL import Image, ImageFont, ImageDraw

AllowedActions = ['both', 'publish', 'subscribe']

# Configure logging
logger = logging.getLogger("AWSIoTPythonSDK.core")
logger.setLevel(logging.DEBUG)
streamHandler = logging.StreamHandler()
formatter = logging.Formatter('%(asctime)s - %(name)s - %(levelname)s - %(message)s')
streamHandler.setFormatter(formatter)
logger.addHandler(streamHandler)


def process_message(client, userdata, message):
    if message.topic == topic:
        message_payload = json.loads(message.payload)
        status_type = message_payload['type']
        if status_type == 'image':
            set_eink_image(status=message_payload['data'])


def set_eink_image(status: str = 'idle', path: str = "img"):
    filename = f"{status}.png"
    print(f'Using filename: {filename}')
    inky_display = auto(ask_user=True, verbose=True)
    inky_display.set_border(BLACK)

    img = Image.open(os.path.join(path, filename))
    # inky_display.set_image(img.rotate(180))
    inky_display.set_image(img)
    inky_display.show()


config = configparser.ConfigParser()
config.read('config.ini')

host = config['aws_iot'].get('endpoint')
rootCAPath = config['certs']['root_ca']
certificatePath = config['certs']['iot_cert']
privateKeyPath = config['certs']['iot_priv_key']
port = config['aws_iot'].get('port') or None
useWebsocket = config['aws_iot'].get('port') or False
clientId = config['aws_iot'].get('client_id') or 'StatusBoard'
topic = config['aws_iot'].get('topic') or 'status_board'
mode = config['aws_iot'].get('mode') or 'both'
defaultStatus = config['status_board'].get('default_status') or None

if mode not in AllowedActions:
    logger.error("Unknown --mode option %s. Must be one of %s" % (mode, str(AllowedActions)))
    exit(2)

if useWebsocket and certificatePath and privateKeyPath:
    logger.error("X.509 cert authentication and WebSocket are mutual exclusive. Please pick one.")
    exit(2)

if not useWebsocket and (not certificatePath or not privateKeyPath):
    logger.error("Missing credentials for authentication.")
    exit(2)

# Port defaults
if useWebsocket and not port:  # When no port override for WebSocket, default to 443
    port = 443
if not useWebsocket and not port:  # When no port override for non-WebSocket, default to 8883
    port = 8883

# Init AWSIoTMQTTClient
myAWSIoTMQTTClient = None
if useWebsocket:
    myAWSIoTMQTTClient = AWSIoTMQTTClient(clientId, useWebsocket=True)
    myAWSIoTMQTTClient.configureEndpoint(host, port)
    myAWSIoTMQTTClient.configureCredentials(rootCAPath)
else:
    myAWSIoTMQTTClient = AWSIoTMQTTClient(clientId)
    myAWSIoTMQTTClient.configureEndpoint(host, port)
    myAWSIoTMQTTClient.configureCredentials(rootCAPath, privateKeyPath, certificatePath)

# AWSIoTMQTTClient connection configuration
myAWSIoTMQTTClient.configureAutoReconnectBackoffTime(1, 32, 20)
myAWSIoTMQTTClient.configureOfflinePublishQueueing(-1)  # Infinite offline Publish queueing
myAWSIoTMQTTClient.configureDrainingFrequency(2)  # Draining: 2 Hz
myAWSIoTMQTTClient.configureConnectDisconnectTimeout(10)  # 10 sec
myAWSIoTMQTTClient.configureMQTTOperationTimeout(5)  # 5 sec

# Connect and subscribe to AWS IoT
myAWSIoTMQTTClient.connect()
if mode == 'both' or mode == 'subscribe':
    myAWSIoTMQTTClient.subscribe(topic, 1, process_message)
time.sleep(2)

# Publish to the same topic in a loop forever
loopCount = 0
while True:
    time.sleep(5)


psb/main.py
"""Entry point for psb.
"""
import json
import time

from AWSIoTPythonSDK.MQTTLib import AWSIoTMQTTClient

from psb import logger
from psb.conf import PsbConfig
from psb.display import EinkDisplay


def process_message(client, userdata, message): # noqa
    config = PsbConfig()
    eink_display = EinkDisplay()
    logger.info('Processing incoming message')
    if message.topic == config.topic:
        message_payload = json.loads(message.payload)
        status_type = message_payload['type']
        if status_type == 'image':
            logger.info('Setting eink display to image')
            eink_display.image(status=message_payload['data'], path=config.img_location)
        elif status_type == 'text':
            logger.info('Setting eink display to text')
            eink_display.text(text=message_payload['data'])
    else:
        logger.info(f'Ignoring message for topic {message.topic}')


def shell():
    """Command line 'psb'.
    """
    config = PsbConfig()

    # Init AWSIoTMQTTClient
    mqtt_client = None
    if config.use_websockets:
        mqtt_client = AWSIoTMQTTClient(config.client_id, useWebsocket=True)
        mqtt_client.configureEndpoint(config.endpoint, config.port)
        mqtt_client.configureCredentials(config.root_ca)
    else:
        mqtt_client = AWSIoTMQTTClient(config.client_id)
        mqtt_client.configureEndpoint(config.endpoint, config.port)
        mqtt_client.configureCredentials(config.root_ca, config.aws_iot_priv_key, config.aws_iot_cert)

    # AWSIoTMQTTClient connection configuration
    mqtt_client.configureAutoReconnectBackoffTime(1, 32, 20)
    mqtt_client.configureOfflinePublishQueueing(-1)  # Infinite offline Publish queueing
    mqtt_client.configureDrainingFrequency(2)  # Draining: 2 Hz
    mqtt_client.configureConnectDisconnectTimeout(10)  # 10 sec
    mqtt_client.configureMQTTOperationTimeout(5)  # 5 sec

    # Connect and subscribe to AWS IoT
    mqtt_client.connect()
    if config.mode == 'both' or config.mode == 'subscribe':
        mqtt_client.subscribe(config.topic, 1, process_message)
    time.sleep(2)

    # Publish to the same topic in a loop forever
    while True:
        time.sleep(5)


psb/display.py
"""Handler for the image/text processing and display to eInk."""
import os

from font_fredoka_one import FredokaOne # noqa
from inky import BLACK, InkyPHAT
from inky.auto import auto
from PIL import Image, ImageDraw, ImageFont

from psb import logger


class EinkDisplay:
    """Object controlling the display of content to the eink display.
    """
    def __init__(self):
        """Init function for EinkDisplay.
        """
        try:
            self.inky_display = auto(ask_user=False, verbose=True)
        except RuntimeError:
            logger.error('Problem connecting to eInk display automatically, trying manually.')
            self.inky_display = InkyPHAT('black')

        self.inky_display.set_border(BLACK)

    @staticmethod
    def process_message(text: str) -> list:
        """Static method to process incoming message.

        Parameters
        ----------
        text : str
            The contents of message to be processed.

        Returns
        -------
        list
            The message after being processed.
        """
        logger.info('Message requires processing.')
        split_lines = []
        lines = text.split("\n")
        for line in lines:
            if len(line) > 20:
                logger.info(f'Line, \"{line}\" longer than 20 characters, splitting')
                line_split = line.split(' ')
                logger.debug(f'Line when split: {line_split}')
                new_line = line_split.pop(0)
                logger.info(new_line)
                while line_split:
                    logger.debug(f'Contents of left in line_split: {line_split}')
                    logger.debug('Line length after join: %s' % len(" ".join([new_line, line_split[0]])))
                    if len(" ".join([new_line, line_split[0]])) <= 20:
                        new_line = " ".join([new_line, line_split.pop(0)])
                    else:
                        split_lines.append(new_line)
                        new_line = line_split.pop(0)
                split_lines.append(new_line)
            else:
                logger.info(f'Line, \"{line}\" does not require splitting.')
                split_lines.append(line)
        logger.debug(f'Contents of split lines: {split_lines}')
        return split_lines

    def image(self, path: str, status: str = 'idle'):
        """Function to process the request for an image to be displayed.

        Parameters
        ----------
        path : str
            The path that locates the image files.
        status : str, optional
            The state (and corrosponding filename) use to display, by default 'idle'
        """
        filename = f"{status}.png"
        logger.info(f'Using filename: {filename}')
        try:
            img = Image.open(os.path.join(path, filename))
            self.__eink_show(img)
        except FileNotFoundError as file_not_found:
            logger.error(f'Cannot set image, no image {filename}')
            raise FileNotFoundError from file_not_found

    def text(self, text: str):
        """Function to process request for text to be displayed on the eink display.
        Will check that there are no more than 3 lines of 20 characters each, if so
        they will be cut down to to correct size. If more than 3 lines, no text is displayed.txt.

        Parameters
        ----------
        text : str
            The text string to process.
        """
        logger.info(f'Setting display to text: {text}')
        img = Image.new("P", (self.inky_display.WIDTH, self.inky_display.HEIGHT))
        draw = ImageDraw.Draw(img)
        font = ImageFont.truetype(FredokaOne, 22)
        x_axis = 0
        y_axis = 0

        if len(text) <= 20 and len(text.split("\n")) == 1: # noqa #TODO: fix the amount of if statements.
            logger.info('Message 1 line and <= 20 characters, displaying.')
            draw.text((x_axis, y_axis), text, self.inky_display.BLACK, font)
        else:
            processed_message = self.process_message(text)
            if len(processed_message) <= 3:
                text = "\n".join(processed_message)
                logger.info(f'New message: {text}')
                draw.multiline_text((x_axis, y_axis), text, self.inky_display.BLACK, font, align="center")
                self.__eink_show(img)
            else:
                logger.error(f'Cannot display message, more than 3 lines long, ({len(processed_message)})')
                logger.error(f'Post-processing message: {processed_message}')

    def __eink_show(self, img: Image):
        """Function to show completed image on display.png

        Parameters
        ----------
        img : Image
            The PIL image to display.
        """
        self.inky_display.set_image(img)
        self.inky_display.show()


status_board.py
from AWSIoTPythonSDK.MQTTLib import AWSIoTMQTTClient
import logging
import time
import json
import configparser
import os
from inky import BLACK
from inky.auto import auto
from font_fredoka_one import FredokaOne
from PIL import Image, ImageFont, ImageDraw
from pathlib import Path

AllowedActions = ['both', 'publish', 'subscribe']

# Configure logging
logger = logging.getLogger()
logger.setLevel(logging.INFO)
streamHandler = logging.StreamHandler()
formatter = logging.Formatter('%(asctime)s - %(name)s - %(levelname)s - %(message)s')
streamHandler.setFormatter(formatter)
logger.addHandler(streamHandler)

base_path = str(Path(__file__).absolute().parent)
img_path = base_path + "resources/img"


def process_message(client, userdata, message):
    logger.info('Processing incoming message')
    if message.topic == topic:
        message_payload = json.loads(message.payload)
        status_type = message_payload['type']
        if status_type == 'image':
            logger.info('Setting eink display to image')
            set_eink_image(status=message_payload['data'], path=img_path)
        elif status_type == 'text':
            logger.info('Setting eink display to text')
            set_eink_text()
    else:
        logger.info(f'Ignoring message for topic {topic}')


def set_eink_image(path: str, status: str = 'idle'):
    filename = f"{status}.png"
    logger.info(f'Using filename: {filename}')
    inky_display = auto(ask_user=True, verbose=True)
    inky_display.set_border(BLACK)

    img = Image.open(os.path.join(path, filename))
    # inky_display.set_image(img.rotate(180))
    inky_display.set_image(img)
    inky_display.show()


def set_eink_text():
    inky_display = auto(ask_user=True, verbose=True)
    inky_display.set_border(BLACK)
    img = Image.new("P", (inky_display.WIDTH, inky_display.HEIGHT))
    draw = ImageDraw.Draw(img)
    font = ImageFont.truetype(FredokaOne, 22)
    message = "Currently playing\nFall Out New Vegas"
    w, h = font.getsize(message)
    x = 0
    y = 0
    # draw.text((x, y), message, inky_display.BLACK, font)
    draw.multiline_text((x, y), message, inky_display.BLACK, font, align="center")
    # inky_display.set_image(img.rotate(180))
    inky_display.set_image(img)
    inky_display.show()


config = configparser.ConfigParser()
config.read(base_path + "/" + 'config.ini')

if Path(config['certs']['root_ca']).is_absolute():
    rootCAPath = config['certs']['root_ca']
else:
    rootCAPath = base_path + "/" + config['certs']['root_ca']

if Path(config['certs']['iot_cert']).is_absolute():
    certificatePath = config['certs']['iot_cert']
else:
    certificatePath = base_path + "/" + config['certs']['iot_cert']

if Path(config['certs']['iot_priv_key']).is_absolute():
    privateKeyPath = config['certs']['iot_priv_key']
else:
    privateKeyPath = base_path + "/" + config['certs']['iot_priv_key']

host = config['aws_iot'].get('endpoint')
port = config['aws_iot'].get('port') or None
useWebsocket = config['aws_iot'].get('port') or False
clientId = config['aws_iot'].get('client_id') or 'StatusBoard'
topic = config['aws_iot'].get('topic') or 'status_board'
mode = config['aws_iot'].get('mode') or 'both'
defaultStatus = config['status_board'].get('default_status') or None

logger.info(f'Using AWS root certificate: {rootCAPath}')
logger.info(f'Using AWS IoT Thing certificate: {certificatePath}')
logger.info(f'Using AWS IoT Thing private key: {privateKeyPath}')
logger.info(f'Using endpoint {host}:{port}')
logger.info(f'Using topic {topic}')

if mode not in AllowedActions:
    logger.error("Unknown --mode option %s. Must be one of %s" % (mode, str(AllowedActions)))
    exit(2)

if useWebsocket and certificatePath and privateKeyPath:
    logger.error("X.509 cert authentication and WebSocket are mutual exclusive. Please pick one.")
    exit(2)

if not useWebsocket and (not certificatePath or not privateKeyPath):
    logger.error("Missing credentials for authentication.")
    exit(2)

# Port defaults
if useWebsocket and not port:  # When no port override for WebSocket, default to 443
    port = 443
if not useWebsocket and not port:  # When no port override for non-WebSocket, default to 8883
    port = 8883

# Init AWSIoTMQTTClient
myAWSIoTMQTTClient = None
if useWebsocket:
    myAWSIoTMQTTClient = AWSIoTMQTTClient(clientId, useWebsocket=True)
    myAWSIoTMQTTClient.configureEndpoint(host, port)
    myAWSIoTMQTTClient.configureCredentials(rootCAPath)
else:
    myAWSIoTMQTTClient = AWSIoTMQTTClient(clientId)
    myAWSIoTMQTTClient.configureEndpoint(host, port)
    myAWSIoTMQTTClient.configureCredentials(rootCAPath, privateKeyPath, certificatePath)

# AWSIoTMQTTClient connection configuration
myAWSIoTMQTTClient.configureAutoReconnectBackoffTime(1, 32, 20)
myAWSIoTMQTTClient.configureOfflinePublishQueueing(-1)  # Infinite offline Publish queueing
myAWSIoTMQTTClient.configureDrainingFrequency(2)  # Draining: 2 Hz
myAWSIoTMQTTClient.configureConnectDisconnectTimeout(10)  # 10 sec
myAWSIoTMQTTClient.configureMQTTOperationTimeout(5)  # 5 sec

# Connect and subscribe to AWS IoT
myAWSIoTMQTTClient.connect()
if mode == 'both' or mode == 'subscribe':
    myAWSIoTMQTTClient.subscribe(topic, 1, process_message)
time.sleep(2)

# Publish to the same topic in a loop forever
loopCount = 0
while True:
    time.sleep(5)
